src/os/bluestore/simple_bitmap.h

```diff
@@ -0,0 +1,139 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+#pragma once
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <cstring>
+#include <cmath>
+#include <iomanip>
+
+#include "include/ceph_assert.h"
+
+struct extent_t {
+  uint64_t offset;
+  uint64_t length;
+};
+
+class SimpleBitmap {
+public:
+  //----------------------------------------------------------------------------
+  friend std::ostream& operator<<(std::ostream& os, const SimpleBitmap& bmap);
+
+  //----------------------------------------------------------------------------
+  //throw bad_alloc
+  SimpleBitmap(CephContext *_cct, const std::string &_path, uint64_t num_bits);
+
+  //----------------------------------------------------------------------------
+  ~SimpleBitmap();
+
+  // set a bit range range of @length starting at @offset
+  bool     set(uint64_t offset, uint64_t length);
+  // clear a bit range range of @length starting at @offset
+  bool     clr(uint64_t offset, uint64_t length);
+
+  // returns a copy of the next set extent starting at @offset
+  extent_t get_next_set_extent(uint64_t offset);
+
+  // returns a copy of the next clear extent starting at @offset
+  extent_t get_next_clr_extent(uint64_t offset);
+
+  //----------------------------------------------------------------------------
+  inline uint64_t get_size() {
+    return m_num_bits;
+  }
+
+  //----------------------------------------------------------------------------
+  // clears all bits in the bitmap
+  inline void clear_all() {
+    std::memset(m_arr, 0, words_to_bytes(m_word_count));
+  }
+
+  //----------------------------------------------------------------------------
+  // sets all bits in the bitmap
+  inline void set_all() {
+    std::memset(m_arr, 0xFF,  words_to_bytes(m_word_count));
+  }
+
+  //----------------------------------------------------------------------------
+  bool bit_is_set(uint64_t offset) {
+    if (offset < m_num_bits) {
+      uint64_t      word_index = offset_to_index(offset);
+      uint64_t      bit_offset = (offset & BITS_IN_WORD_MASK);
+      uint64_t mask       = 1ULL << bit_offset;
+      return (m_arr[word_index] & mask);
+    } else {
+      return false;
+    }
+  }
+
+  //----------------------------------------------------------------------------
+  bool bit_is_clr(uint64_t offset) {
+    if (offset < m_num_bits) {
+      uint64_t      word_index = offset_to_index(offset);
+      uint64_t      bit_offset = (offset & BITS_IN_WORD_MASK);
+      uint64_t mask       = 1ULL << bit_offset;
+      return ( (m_arr[word_index] & mask) == 0 );
+    } else {
+      return false;
+    }
+  }
+
+private:
+  void __set_single_bit_no_bounderies_check(uint64_t offset);
+  void __clr_single_bit_no_bounderies_check(uint64_t offset);
+
+  //---------------------------------------------------------------------------
```

Comment:
How about single function
```
  static inline std::pair<uint64_t, uint64_t> split(uint64_t offset) {
    return { offset >> BITS_IN_WORD_SHIFT, offset & BITS_IN_WORD_MASK };
  }
```
that can be used as:
`auto [word_index, bit_index ] = split(offset);` ?

It is still the same performance, but looks nicer.